File: src/FXRateDashboard/Services/RateQueryMapper.cs
using FXRateDashboard.Models;

namespace FXRateDashboard.Services;

public sealed class RateQueryMapper : IRateQueryMapper
{
    public RateHistoryQuery MapHistoryQuery(TimeRangePreset range, DateTimeOffset? nowUtc = null)
    {
        var anchor = (nowUtc ?? DateTimeOffset.UtcNow).ToUniversalTime();

        return range switch
        {
            TimeRangePreset.Day => new RateHistoryQuery(range, anchor.AddDays(-1), anchor, "minute"),
            TimeRangePreset.Week => new RateHistoryQuery(range, anchor.AddDays(-7), anchor, "hour"),
            TimeRangePreset.Month => new RateHistoryQuery(range, anchor.AddDays(-30), anchor, "day"),
            TimeRangePreset.Year => new RateHistoryQuery(range, anchor.AddDays(-365), anchor, "day"),
            _ => throw new ArgumentOutOfRangeException(nameof(range), range, null)
        };
    }
}

File: macos/FXRateDashboard/Sources/FXRateDashboardKit/Services/RateQueryMapper.swift
import Foundation

public protocol RateQueryMapperProtocol: Sendable {
    func mapHistoryQuery(_ range: TimeRangePreset, nowUTC: Date) -> RateHistoryQuery
}

public struct RateQueryMapper: RateQueryMapperProtocol {
    public init() {}

    public func mapHistoryQuery(_ range: TimeRangePreset, nowUTC: Date = Date()) -> RateHistoryQuery {
        let anchor = nowUTC

        return switch range {
        case .day:
            RateHistoryQuery(range: range, fromUTC: anchor.addingTimeInterval(-86_400), toUTC: anchor, group: "minute")
        case .week:
            RateHistoryQuery(range: range, fromUTC: anchor.addingTimeInterval(-604_800), toUTC: anchor, group: "hour")
        case .month:
            RateHistoryQuery(range: range, fromUTC: anchor.addingTimeInterval(-2_592_000), toUTC: anchor, group: "day")
        case .year:
            RateHistoryQuery(range: range, fromUTC: anchor.addingTimeInterval(-31_536_000), toUTC: anchor, group: "day")
        }
    }
}
